Stationary teleporters

TheShuEd opened the pull request:
## About the PR
A new Gateways option has been added as an alternative system. 
![image](https://github.com/user-attachments/assets/92cd92eb-8552-498f-859e-df267ce11a6f)
<img width="706" height="340" alt="image" src="https://github.com/user-attachments/assets/4b48dd86-45ca-43b3-80de-c533957a149e" />

This system is implied to be player-controlled. It comes along with several new elements:
- Gateways chips: Data carriers of gateway coordinates. Disposable: they can be used to record coordinates, but cannot be erased. In the future could be used as loot for Salvage, with planet coordinates. 
- Gateways control console: A console into which you can insert chips with Gateways coordinates. Once the chips are inserted, the console can be used to link and control the routes of the Gateways in any way. Locked to Command Access.
- Autolinking for chips and gateways. Chips and Gateways with the same key are automatically linked when the map is initialized, which makes it convenient to use them for mapping

At the moment, neither the console nor the chips can be obtained by players, but after testing, it could be T3 technology.
Such a gateway system is emergent enough that mechanics could be added to it.
Possible potential scenarios:
- Traitors steal chips to destabilize the portals.
- Nuclear operatives secretly read the coordinates of the gateways, and directly break into the center of the station through them. (Gate-ops?)

## Why / Balance
Primarily as the core mechanics for the Gate map #32032
With the option to develop a Salvage Proposal in future 


## Basic Testing plan

- Create 2 `StationTeleporter`, `ComputerStationTeleportersControl`, and 2 `TeleporterChipBlank` (basic entities created by players)
- Click on different teleporters with the chips to link their coordinates.
- Place the chips with linked coordinates inside the console.
- Open the console (commander access required) and verify that both portals are now visible inside the console and can be linked to each other.
- Optionally, you can place the chips in `ComputerSyndicateTeleportersControl` and open the portals to ensure they are red.
- You can also verify that turning off the power shuts down the portals, and that they are restored once the power is turned back on.

### Interacting with HandTeleporter

- Create 2 portals using `HandTeleporter`
- Place the teleporter inside `ComputerStationTeleportersControl`
- Verify that the console displays the portals’ locations, that they are linked, and that they can be relinked to other portals.

### Testing Auto Link for mapping.

- Create any number of teleporters from different sections, such as `StationTeleporterSalvage`
- Create a `StationTeleporter` that does not have auto-linking enabled.
- Create a `ComputerStationTeleportersControlFilled` console.
- Verify that the console is populated with chips and linked to the teleporters created earlier, except for the teleporter without auto-linking.

## Media
![image](https://github.com/user-attachments/assets/ca3c9e53-ee84-4d5a-9be9-33a3da119433)

https://github.com/user-attachments/assets/b346aa72-e05a-4225-aedf-57f3c5239591

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

:cl:
- add: New T3 “TeleSci” technology added. Allows you to create a portal control console, a stationary teleporter, and coordinate chips.

TheShuEd committed:
sprite, start work on systems

TheShuEd committed:
Woah! Its work!

TheShuEd committed:
Show gateways! Yay!

TheShuEd committed:
Update ring.png

TheShuEd committed:
Update portal.yml

TheShuEd committed:
bububui

TheShuEd committed:
Fuck UI

TheShuEd committed:
draw lines!

TheShuEd committed:
clear old links lines

TheShuEd committed:
UI button work now!

TheShuEd committed:
we can cut links via UI now!

TheShuEd committed:
yay! linking via UI

TheShuEd committed:
comfortable linking

TheShuEd committed:
Update hand_teleporter.yml

TheShuEd committed:
ambience, linked entities events

TheShuEd committed:
nice connect and disconnect sounds!

TheShuEd committed:
unpowering and memory

TheShuEd committed:
power fix

TheShuEd committed:
chip added

TheShuEd committed:
chip linking & auto linking

TheShuEd committed:
chip examined

TheShuEd committed:
auto linked chips&gates variants

TheShuEd commented:
It's a bit crooked-drafted still, but I need integration tests.

TheShuEd committed:
Update meta.json

TheShuEd commented:
TODO:
- [ ] Fix display of “links” if we don't know the coordinates of the second portal
- [ ] Different icons on the map for linked and unlinked portals (or colors)
- [ ] Attempt to add the ability to link portals via clicks on NavMap blips
- [ ]

TheShuEd committed:
Update computers.yml

mirrorcult commented:
"gateways", at least in ss13, refers to the station gateways that take you to away missions / other maps, i think calling these "gateways" (especially since the entities we have use the same sprites) might be confusing, especially since theres been some consideration towards porting some variation of the ss13 concept

mirrorcult commented:
at least i personally think it can be confusing if "gateways" both refers to always-available things that let you teleport around the station, as well as a separate conditionally-available thing that lets you teleport down to planetmaps and stuff

TheShuEd commented:
I haven't played ss13, but here's what I think:
- It has been said more than once that we are not making a complete copy of ss13, and at the moment we already have a huge difference in gameplay. The process of adaptation to the new gateways format will take 1-2 interactions with the mechanics, which is not a problem
- If we plan to have both station-based stable teleports and planet-based teleports, I would prefer that they be identical systems operating on similar principles rather than two completely different ones.
- gateways on stations shouldn't be permanent, something players can see. I'd either lock them behind T3, or leave them purely on some maps as gimmicks

TheShuEd commented:
well and the current implementation will allow using gateways only for salvage if we abandon the idea of station gates. nevertheless, it CAN be used in this format for different scenarios, which has a good effect on the sandbox component of the game

TheShuEd committed:
syndicate gateway console, portal coloring, realtime UI update

TheShuEd committed:
QoL update

TheShuEd commented:
https://github.com/user-attachments/assets/a700633e-6614-45b5-8913-9150ba2f316f

QoL update:
Ability to interact with portals on the map itself, visualization of unpowered and linked portals

TheShuEd committed:
Circuitboards

TheShuEd committed:
Update StationGatewayNavMapControl.cs

TheShuEd committed:
optimize Draw(), add Filled console variant

TheShuEd commented:
There are some issues with gate connection line drawing that may randomly appear in a round, or they may be clearly visible on the Dev map. But I have no idea how to fix them.

TheShuEd commented:
https://github.com/user-attachments/assets/7d73dc6a-4187-4f5c-b89e-f79323f04c47

Train map linklines also cursed :(

TheShuEd committed:
documentation

TheShuEd committed:
Merge branch 'master' into ed-11-09-2024-station-gates

Fildrance commented:
> at least i personally think it can be confusing if "gateways" both refers to always-available things that let you teleport around the station, as well as a separate conditionally-available thing that lets you teleport down to planetmaps and stuff

As Ed said - it is part of gateway planned evolution. Currently console will be able to control to which it is connected inside one map. It will be leading to 'missions' in future also. Maybe with some restrictions - some gateways will work only internally, while other (single one for station) will lead to mission OR to other internal gateway. New sprite for **internal gateway** explicitly could be nice, but we have gateways working as internal portals already on Train station, so there won't be any more confusion that we already have - so new sprite for 'NT gateway' can wait.

TheShuEd committed:
Merge branch 'master' into ed-11-09-2024-station-gates

KaiserMaus commented:
Cool

TheShuEd committed:
Merge branch 'master' into ed-11-09-2024-station-gates

TheShuEd committed:
mass renaming bro

TheShuEd committed:
resprited

TheShuEd commented:
![image](https://github.com/user-attachments/assets/0fe21d21-8836-429a-a62e-522b336a1ec9)
based on project managers feedback
resprited and renamed to teleporters

TheShuEd committed:
resprite agani

TheShuEd commented:
![image](https://github.com/user-attachments/assets/485b3852-91d7-4ff2-9d40-fc59c4137710)
another resprite

slarticodefast commented on Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml:
```diff
@@ -415,3 +415,23 @@
+  parent: BaseComputerCircuitboard
+  id: SyndicateTeleportersControlCircuitboard
+  name: syndicate teleporters control console board
+  description: A computer printed circuit board for a teleporter controlling console.
```
```suggestion
  description: A computer printed circuit board for a teleporter control console.
```

slarticodefast commented on Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml:
```diff
@@ -415,3 +415,23 @@
+  id: StationTeleportersControlCircuitboard
+  name: teleporters control console board
+  description: A computer printed circuit board for a teleporter controlling console.
+  components:
+    - type: Sprite
+      state: cpu_command
+    - type: ComputerBoard
+      prototype: ComputerStationTeleportersControl
```
```suggestion
  components:
  - type: Sprite
    state: cpu_command
  - type: ComputerBoard
    prototype: ComputerStationTeleportersControl
```
indentation

slarticodefast commented on Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml:
```diff
@@ -415,3 +415,23 @@
+  id: SyndicateTeleportersControlCircuitboard
+  name: syndicate teleporters control console board
+  description: A computer printed circuit board for a teleporter controlling console.
+  components:
+    - type: ComputerBoard
+      prototype: ComputerSyndicateTeleportersControl
```
```suggestion
  components:
  - type: ComputerBoard
    prototype: ComputerSyndicateTeleportersControl
```

slarticodefast commented on Resources/Prototypes/Entities/Objects/Devices/teleporters_chips.yml:
```diff
@@ -0,0 +1,108 @@
+  parent: BaseItem
+  suffix: Blank
+  name: teleporter chip
+  description: A chip with stored teleporter coordinates. Insert it into the teleporter control console to be able to control it.
```
```suggestion
  description: A chip that can store teleporter coordinates. Insert it into the teleporter control console to be able to control the linked portal.
```

slarticodefast commented on Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml:
```diff
@@ -457,6 +457,107 @@
+      state: generic_keyboard
+    - map: ["computerLayerScreen"]
+      state: teleporters_station
+    - map: ["computerLayerKeys"]
+      state: telesci_key
```
```suggestion
    - map: ["computerLayerKeys"]
      state: telesci_key
    - map: [ "enum.WiresVisualLayers.MaintenancePanel" ]
      state: generic_panel_open
```
wire panel is missing

slarticodefast commented on Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml:
```diff
@@ -457,6 +457,107 @@
+    interfaces:
+      enum.StationTeleporterUIKey.Key:
+        type: StationTeleporterBoundUserInterface
+      enum.StorageUiKey.Key:
+        type: StorageBoundUserInterface
```
```suggestion
      enum.StorageUiKey.Key:
        type: StorageBoundUserInterface
      enum.WiresUiKey.Key:
        type: WiresBoundUserInterface
```

slarticodefast commented on Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml:
```diff
@@ -457,6 +457,107 @@
+      state: generic_keyboard
+    - map: ["computerLayerScreen"]
+      state: teleporters_syndicate
+    - map: ["computerLayerKeys"]
+      state: syndie_key
```
```suggestion
    - map: ["computerLayerKeys"]
      state: syndie_key
    - map: [ "enum.WiresVisualLayers.MaintenancePanel" ]
      state: generic_panel_open
```

slarticodefast commented on Resources/Prototypes/Entities/Structures/Machines/station_teleporter.yml:
```diff
@@ -0,0 +1,129 @@
+  parent: [ BaseMachinePowered ]
+  id: StationTeleporter
+  name: stationary teleporter
+  description: The peak of bluespace technology. Giant teleporter that allow you to travel vast distances in the blink of an eye.
```
```suggestion
  description: The peak of bluespace technology. A giant teleporter that allow you to travel vast distances in the blink of an eye.
```

slarticodefast commented on Resources/Prototypes/tags.yml:
```diff
@@ -584,6 +584,9 @@
 - type: Tag
   id: GasVent
 
+- type: Tag
+  id: TeleporterChip
+
```
Put it in alphabetical order!

slarticodefast reviewed (dismissed):
Ok, some stuff I found so far. The console UI needs prediction.
Also I think the way the coordinates are sent to the NavMap can probably be improved as you are converting between different coordinate types a few times, but I will have to take a closer look for that.
Maybe @chromiumboy can help here as well since he made the atmos monitoring console, which is similar.

TheShuEd committed:
Update tags.yml

TheShuEd committed:
Update Resources/Prototypes/Entities/Structures/Machines/station_teleporter.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Objects/Devices/teleporters_chips.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Content.Server/StationTeleporter/Components/TeleporterChipComponent.cs

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
Update Content.Shared/StationTeleporter/StationTeleporterComponent.cs

Co-authored-by: slarticodefast <161409025+slarticodefast@users.noreply.github.com>

TheShuEd committed:
renaming renaming

TheShuEd committed:
sharedization

TheShuEd committed:
UI fix

TheShuEd committed:
Data simplify

TheShuEd commented on Content.Server/StationTeleporter/Systems/StationTeleporterSystem.cs:
```diff
@@ -0,0 +1,249 @@
+            break;
+        }
+        if (!successLink && chip.Comp.DeleteOnFailedLink)
+            QueueDel(chip);
```
I need to delete it if it's gone through all the teleporters and hasn't found one that fits.
I'm not quite sure how to avoid the bool variable here

TheShuEd committed:
remove autodeletion

TheShuEd committed:
PostWallDrawingAction

TheShuEd committed:
Update TeleporterChipComponent.cs

TheShuEd commented:
Github Copilot PR summary: (I wanna test it)

This pull request introduces a new feature for managing teleporters via a console interface in the `Content.Client` and `Content.Shared` projects. The changes include the addition of new components, user interface elements, and system logic to support the functionality. The most important changes are summarized below:

### New User Interface and Components for Teleporters:

* [`Content.Client/StationTeleporter/StationTeleporterConsoleBoundUserInterface.cs`](diffhunk://#diff-1d752a68eb2c84565c7ea60b83237d37abb6eb17a71cb5e8e90896845c2055b3R1-R51): Added a new class to handle the user interface logic for the teleporter console. This includes methods for opening the interface, updating its state, and sending messages.
* [`Content.Client/StationTeleporter/StationTeleporterConsoleWindow.xaml`](diffhunk://#diff-33a60488497d47e85e96202b6f5efb6b7584ae3b9103f6a2c8eec373f41b2090R1-R50): Created a new XAML file defining the layout and elements of the teleporter console window, including a map control and a list of teleporters.
* [`Content.Client/StationTeleporter/StationTeleporterConsoleWindow.xaml.cs`](diffhunk://#diff-15d17767c95046c877b461fd96253cb383f958f2980930aa3d90b7432563a70bR1-R284): Implemented the code-behind for the teleporter console window, handling user interactions and updating the display of teleporters.

### Teleporter System Logic:

* [`Content.Client/StationTeleporter/StationTeleporterSystem.cs`](diffhunk://#diff-f0938dafaef5c88b54411ca11bfeed8b1b26d472efa9ef6d590dd1b9d6010dadR1-R34): Added a new system to handle appearance changes for teleporters, updating their visual representation based on their state.
* [`Content.Server/StationTeleporter/Systems/StationTeleporterSystem.cs`](diffhunk://#diff-25daf1c294d9e6bd708f38b694c362611a60ef84baa0353330e655b80e909ce5R1-R40): Implemented server-side logic for initializing teleporter chips and linking them to teleporters based on an auto-link key.

### Shared Components for Teleporters:

* [`Content.Shared/StationTeleporter/Components/StationTeleporterComponent.cs`](diffhunk://#diff-4e34f06239cb65bbd7a20fdd630b81b3cb46f671065e146f21bf839022208d24R1-R54): Defined a new component to represent teleporters, including fields for their name, auto-link key, and sounds for linking and unlinking.
* [`Content.Shared/StationTeleporter/Components/StationTeleporterConsoleComponent.cs`](diffhunk://#diff-c321c172bbda343604b01fac618314f638f3461fa4a4ae84616931a4b468ac11R1-R36): Created a component for the teleporter console, including fields for the selected teleporter, portal color, and chip storage.
* [`Content.Shared/StationTeleporter/Components/TeleporterChipComponent.cs`](diffhunk://#diff-5565286fa6d86d2cfc7c989ebd8127a86165996d183615bb9c5bbdb1236354aaR1-R39): Added a component for teleporter chips, including fields for the connected teleporter, connected name, and auto-link key.

These changes collectively provide a comprehensive system for managing teleporters through a user interface, enhancing the gameplay experience by allowing players to interact with and control teleporters in a more intuitive manner.

TheShuEd committed:
fix lines offset

TheShuEd committed:
labeler renaming, rebound restiction removed

TheShuEd committed:
prevent link popup and sound spamming

TheShuEd committed:
auto linking refactor

TheShuEd committed:
RD handheld teleporter integration

chromiumboy commented:
This looks really cool! I haven't reviewed any code as yet, but I have some notes on the UI

- The linkage lines between the teleports could use some thickening. I'd draw a rectangle rather than using a line
- The separate item storage UI seems unnecessary. Just have it so that clicking an item on the console inserts it and add a small eject button on the teleporter entry, on the same line as its name, to remove it ([example eject button](https://www.shutterstock.com/image-vector/eject-icon-symbol-vector-illustration-1515096281))
- Increase the width of the teleporter entries to fill the available space on the side panel
- You can avoid the need for a separate locate button by jumping to the teleporter when clicking on its name (this would make it match how the other consoles, like the power and crew monitors, handle it)
- When you click 'start connection' on a teleporter entry, the text on that button should change to 'select a destination' and any valid destinations should have their 'start connection' button change their text to say 'activate connection'. Consider making the 'activate connection' buttons turn green as well when this happens
- Consider renaming 'start connection' to 'create new connection'
- The 'cut connection' button should include the name of the teleporter to which they are connected, e.g. 'cut connection to Cargo'. Also consider coloring them red
- The color palette used for the teleporter entries doesn't match what we typically use for these console

TheShuEd committed:
fix randomteleport bug

TheShuEd commented:
please don't make me redesign the UI, I've suffered so much with it.

Let me do something cool without using UI instead of this.

TheShuEd commented:
like, uh, do you want me to go make a shadow dimension?

TheShuEd committed:
Merge branch 'master' into ed-11-09-2024-station-gates

CRISOOOOO commented:
I have a question, the control console needs to be mapped on the bridge or next to the portal?

TheShuEd commented:
Idk, we dont have standards

chromiumboy committed:
Merge branch 'master' into ed-11-09-2024-station-gates

TheShuEd commented:
Press F to Gate map
Chromium, if you find time to finish it, you can reopen PR

TheShuEd committed:
Merge branch 'master' into ed-11-09-2024-station-gates

TheShuEd committed:
Update hand_teleporter.yml

Fildrance commented:
The fact that it also works with hand-held portal probably should be mentioned in PR description. The way i is interacting too.

Fildrance reviewed (changes requested):
generally cool stuff, minor nitpicks

Fildrance commented:
Having ability to 'clone' data from one chip to other would be nice, as an idea. Probably for future implementation.

TheShuEd committed:
Merge branch 'master' of https://github.com/TheShuEd/ed-station-14 into ed-11-09-2024-station-gates

# Conflicts:
#	Content.Shared/Teleportation/Systems/SharedPortalSystem.cs
#	Resources/Prototypes/Entities/Objects/Devices/Circuitboards/computer.yml
#	Resources/Prototypes/tags.yml
#	Resources/Textures/Structures/Machines/computers.rsi/meta.json

TheShuEd committed:
Merge branch 'master' of https://github.com/TheShuEd/ed-station-14 into ed-11-09-2024-station-gates

# Conflicts:
#	Resources/Prototypes/Entities/Structures/Machines/gateway.yml

TheShuEd committed:
Update SharedStationTeleporterSystem.cs

TheShuEd committed:
Update SharedStationTeleporterSystem.UI.cs

TheShuEd committed:
some clean up

TheShuEd committed:
first smol review pass

TheShuEd committed:
and that

TheShuEd committed:
review pass 2

TheShuEd committed:
Update station_teleporter.yml

TheShuEd committed:
review pass 3

TheShuEd committed:
review pass 4

TheShuEd committed:
finally

TheShuEd committed:
update protos

TheShuEd committed:
another clean up

TheShuEd committed:
fix bugs

TheShuEd committed:
add to players

TheShuEd committed:
remove unnecesary update loop lol.

TheShuEd committed:
Update computers.yml

Fildrance reviewed (changes requested):
Couple of questions and nitpicks.

Flickering bug i mentioned above was fixed but i got other odd thing

https://github.com/user-attachments/assets/b337e0d0-b9c1-4b78-972d-cec5c3c488f1

On top of that - i would like to ask Jez about having this thing as research gameplay-wise, coz its very big change... and even tho we are looking at remaking research related stuff and already merged anomaly reword design - i am not sure how long it would take us to get over place where research points are really well balancer resource! So better safe then sorry...

TheShuEd committed:
fieldrance review apply

TheShuEd committed:
Kazne resprite

Fildrance reviewed (approved):
Got green light design-wise, lets look at how it will play out!
